### PR TITLE
tests: unit tests for workflows/tasks/actions

### PR DESCRIPTION
**Pull request by jacquerie**
- [x] `mark`
- [x] `is_marked`
- [x] `is_record_accepted`
- [x] `shall_halt_workflow`
- [x] `in_production_mode`
- [x] `add_core`
- [x] `halt_record`
- [x] `update_note`
- [x] `reject_record`
- [x] `is_record_relevant`
- [x] `is_experimental_paper`
- [x] `is_arxiv_paper`
- [x] `is_submission`
- [x] `prepare_update_payload`

### CODE REVIEW

**Comment by kaplun** [comment]
@jacquerie great initiative. One important thing: if you see anything that is not obvious or doesn't feel right, always check with me, before writing a test that certify possibly the wrong behavior.
It's the right moment to do a thorough review of the workflow.

File: inspirehep/utils/arxiv.py [hunk 1]
```diff
@@ -24,17 +24,11 @@
...
 def get_clean_arXiv_id(record):
     """Return the arXiv identifier from given record."""
-    arxiv_id = record.get("arxiv_id")
-    if not arxiv_id:
-        arxiv_eprints = record.get('arxiv_eprints', [])
-        for element in arxiv_eprints:
-            if element.get("value", ""):
-                arxiv_id = element.get("value", "")
-
+    arxiv_id = get_value(record, 'arxiv_eprints.value[-1]')
```
Comment (kaplun) [inline inspirehep/utils/arxiv.py]: Use `[0]` rather than `[-1]` as reported in #1751

File: tests/unit/utils/test_utils_arxiv.py [hunk 2]
```diff
@@ -3,63 +3,80 @@
...
+
+def test_get_clean_arXiv_id_from_arxiv_eprints_selects_last():
+    record = {
+        'arxiv_eprints': [
+            {'value': 'oai:arXiv.org:0801.4782'},
+            {'value': 'oai:arXiv.org:0805.1410'},
+        ],
+    }
+
+    expected = '0805.1410'
```
Comment (kaplun) [inline tests/unit/utils/test_utils_arxiv.py]: So here would be the first to be returned.

**Comment by jacquerie** [comment]
> One important thing: if you see anything that is not obvious or doesn't feel right, always check with me, before writing a test that certify possibly the wrong behavior.

Naturally! An example of that was our discussion on file descriptors yesterday in chat.

**Comment by jacquerie** [comment]
BTW you don't have to review this yet, because I'm still discovering things worth fixing in other parts of the code, so commits are going to be duplicated between here and other PRs that I'm spinning off of here.

**Comment by jacquerie** [comment]
`is_arxiv_paper` does not handle this corner case (`arxiv_eprints` with `value`s but without `categories`):
```python
In [1]: class StubObj(object):
   ...:     def __init__(self, data):
   ...:         self.data = data
   ...:

In [2]: from inspirehep.modules.workflows.tasks.actions import is_arxiv_paper

In [3]: obj = StubObj({'arxiv_eprints': [{'value': 'oai:arXiv.org:0801.4782'}]})

In [4]: is_arxiv_paper(obj)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-0f3287181dbc> in <module>()
----> 1 is_arxiv_paper(obj)

/Users/jacquerie/.virtualenvs/inspire/src/inspire-next/inspirehep/modules/workflows/tasks/actions.pyc in is_arxiv_paper(obj, *args, **kwargs)
    159 def is_arxiv_paper(obj, *args, **kwargs):
    160     """Check if the record is from arXiv."""
--> 161     return bool(get_value(obj.data, "arxiv_eprints.categories", [[]])[0]) or \
    162         get_clean_arXiv_id(obj.data)
    163

IndexError: list index out of range
```

**Comment by jacquerie** [comment]
`is_experimental_paper` is completely broken:
```
In [5]: from inspirehep.utils.record import get_value

In [6]: obj = StubObj({'arxiv_eprints': [{'categories': 'hep-ex'}]})

In [7]: list(get_value(obj.data, "arxiv_eprints.categories", [[]])[0])
Out[7]: ['h', 'e', 'p', '-', 'e', 'x']
```

**Comment by kaplun** [comment]
Please send in also the corresponding fixes for the bugs you are finding. Also re: arXiv category, I believe there shouldn't been such case where an arXiv document comes without categories from arXiv.

File: tests/unit/workflows/test_workflows_tasks_arxiv.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+@pytest.fixture
+def another_arxiv_pdf():
+    return pkg_resources.resource_string(
+        __name__, os.path.join('fixtures', '0701031v2.pdf'))
+
+
+class StubObj(object):
+    def __init__(self, data, extra_data, files):
+        self.data = data
+        self.extra_data = extra_data
```
Comment (rikirenz) [inline tests/unit/workflows/test_workflows_tasks_arxiv.py]: `extra_data` is always empty. Why?

Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_arxiv.py]: Mh, you're right, tasks in this file don't use `extra_data` at all.

Comment (jacquerie) [inline tests/unit/workflows/test_workflows_tasks_arxiv.py]: For now, every test file has its own small mocks/stubs/dummies. The better approach would be to write a unique mock that covers all use cases and put it in a `conftest.py` in the same folder.

**Comment by jacquerie** [comment]
> Also re: arXiv category, I believe there shouldn't been such case where an arXiv document comes without categories from arXiv.

But there _could_ be, because we're not enforcing it from the schema. The fact is, the idiom `get_value(record, "foo.bar", [[]])[0]` is incorrect, and it's the common cause of both bugs (but I agree that they differ in severity). So this is going to get fixed anyway : )

**Comment by kaplun** [comment]
The issue is that we could enforce it in the schema, but it's simply that given an arXiv eprint we don't know automatically its categories. Hence one might create a record with just the eprint, and resolve the categories at a later stage. However, upon harvesting we surely know them.

**Comment by jacquerie** [comment]
I see. Then I'm not going to assert that in a test.

**Comment by kaplun** [comment]
I mean, I think it's great to assert in the workflow test that categories *are* there. (since this would make the test to be documentation/specification for the workflow). Instead no need, IMHO, to test for the case, within the workflow, where arXiv doesn't give us categories.

**Comment by jacquerie** [comment]
Minor bug: `is_record_relevant` does not match its signature, because it passes its parameters to `is_submission` which has a fixed number of parameters.

Travis: https://travis-ci.org/inspirehep/inspire-next/jobs/177641066#L1518

**Comment by kaplun** [comment]
@jacquerie what do you do for:
> `is_experimental_paper` is completely broken:

Besides I understand there are no other major mis-behavior you have found right?

**Comment by jacquerie** [comment]
> @jacquerie what do you do for:
>
> is_experimental_paper is completely broken:
>
> Besides I understand there are no other major mis-behavior you have found right?

Oh right! I was wrong about that. Note that I was missing `[]` around `hep-ex`, which is why I was getting that weird result.

**Pull request merged**